### PR TITLE
Fixes #8968 - recover on race condition during fact import

### PR DESCRIPTION
**Pull request by ares**

### CODE REVIEW

**Comment by ares** [comment]
@unorthodoxgeek would you like to review?

**Comment by unorthodoxgeek** [comment]
Sure. First thing tomorrow morning :)

File: app/models/foreman_chef/fact_importer.rb [hunk 1]
```diff
@@ -30,9 +32,21 @@ def add_missing_facts(tree_hash, parent = nil, prefix = '')
...
+          # fact names didn't contain fact_name so we create new one
+          fact_name = fact_name_class.new(:name      => name_with_prefix,
+                                          :parent_id => parent,
+                                          :compose   => compose)
+          if fact_name.save
+            fact_name_id = fact_name.id
+          elsif fact_name.errors[:name].present?
+            # saving could have failed because of raise condition with another import process,
+            # so if the error is on name uniqueness, we try to reload conflicting name and use it
+            conflicting_fact_name = fact_name_class.find_by_name(fact_name.name)
```
Comment (unorthodoxgeek) [inline app/models/foreman_chef/fact_importer.rb]: please use where(:name => fact_name.name).first as a preparation for the rails 4 conventions (the find_by_field methods use method_missing and perform very poorly, they're deprecated)

**Comment by unorthodoxgeek** [comment]
this might still fail if the race is really tight because of transactions.
I think wrapping the method in a transaction would be best. i.e.

```ruby
def add_missing_facts
  fact_name_class.transaction do
    #method
  end
end
```

**Comment by ares** [comment]
Are you sure about the race condition? We only get into the "conflict" branch if fact name save failed (race condition already occured) and the error was on name, if save succeeded, we're ok, in this case, it's another process that will got to elsif branch.

File: app/models/foreman_chef/fact_importer.rb (hunk 1 above)
Comment (ares) [inline app/models/foreman_chef/fact_importer.rb]: ok changed, I'd prefer to use `find_by` in Rails 4 when I got deprecation warning... but using where improves the performance in Rails 3 already

**Comment by unorthodoxgeek** [comment]
yeah, you're probably right, as save is wrapped in transaction anyway and the validations run inside.

**Comment by unorthodoxgeek** [comment]
bottom line is - good to merge :+1:

**Comment by ares** [comment]
great, thanks, merging

**Pull request merged**